src/adjoint.jl

```diff
@@ -1,16 +1,53 @@
 using MacroTools
 using MacroTools: @q, combinedef
+using ChainRulesCore: AbstractZero, Zero, DoesNotExist
 
 named(arg) = isexpr(arg, :(::)) && length(arg.args) == 1 ? :($(gensym())::$(arg.args[1])) : arg
 
 typeless(x) = MacroTools.postwalk(x -> isexpr(x, :(::), :kw) ? x.args[1] : x, x)
 isvararg(x) = isexpr(x, :(::)) && namify(x.args[2]) == :Vararg
 
+function legacytype_warn()
+  # can't use logging macros as that breaks nested AD.
+  Core.println("Zygote internal use  of 'nothing', rather than `AbstractZero`, detected.  This should never occur. Please open an issue on https://github.com/FluxML/Zygote.jl/issues, including the full text of this message. \n Stacktrace:")
+  for (ii, callsite) in enumerate(stacktrace())
+    Core.println("[$ii] $callsite")
+  end
+end
+
+function difftype_error()
```

Comment:
It's unclear to me what's going on in this function based on it's name -- I might have expected to see an error being thrown but that doesn't happen. Calling this `something_error` suggests that it will do something qualitatively different from `legacytype_warn`, but it appears that they do the same thing. Maybe it's better to call both of them `<something>_error` or `<something>_warn` (unless I'm missing something obvious).

Would it be possible to either provide a more descriptive name, or to provide a good comment for the function regarding its intended use / some context?

Reply:
Nice catch, this was also meant to be called `_warn`. I've added some docstrings as well.